File: doctranslator/freezer.py
from __future__ import annotations

import re
from dataclasses import dataclass, field

from doctranslator.sentinels import ANY_SENTINEL_RE, nt_token


_URL = r"https?://[^\s<>()]+"
_EMAIL = r"[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}"
_WIN_PATH = r"(?:[A-Za-z]:\\(?:[^\\/:*?\"<>|\r\n]+\\)*[^\\/:*?\"<>|\r\n]*)"
_PLACEHOLDER = r"(?:\{[^{}\r\n]{1,100}\}|\$\{[^{}\r\n]{1,100}\})"
_PERCENT_SLOT = r"%\d+"
# Freeze legal references as an atomic unit so the translation model cannot drop/alter
# section/article numbers (a common catastrophic failure in contracts).
#
# Examples:
# - Section 7
# - Section 2(c)(i)
# - Sec. 6(b)(iv)
# - Article IV
_EN_LEGAL_REF = (
    r"\b(?:Section|Article|Clause|Paragraph|Schedule)s?\s+"
    r"(?:\d+(?:[.,]\d+)*(?:-\d+(?:[.,]\d+)*)?(?:\([A-Za-z0-9]+\))*|[IVXLCDM]{1,8})\b"
)
_EN_LEGAL_REF_ABBR = (
    r"\b(?:Sec|Art|Cl|Para|Sch)s?\.\s+"
    r"(?:\d+(?:[.,]\d+)*(?:-\d+(?:[.,]\d+)*)?(?:\([A-Za-z0-9]+\))*|[IVXLCDM]{1,8})\b"
)
_ZH_LEGAL_REF = r"第\s*\d+(?:\([A-Za-z0-9]+\))*\s*(?:条|款|项|段|章|节)\b"
_ZH_SCHEDULE_REF = r"(?:附表|附件)\s*\d+(?:\([A-Za-z0-9]+\))*\b"
_CLAUSE_REF = r"\b\d+(?:\([A-Za-z0-9]+\))+"
_ENUM_NUM = r"\(\d{1,3}\)"
_ENUM_ROMAN = r"\((?:[ivxlcdmIVXLCDM]{1,6})\)"
_ENUM_ALPHA = r"\([A-Za-z]\)"
_DOT_LEADER = r"[.\u2026]{8,}"
_UNDERSCORE_LEADER = r"_{5,}"
_DASH_LEADER = r"[-\u2010\u2011\u2012\u2013\u2014\u2015\u2212]{5,}"
_TRADEMARK_TOKEN = r"[A-Za-z0-9]{2,24}[®™℠]"
_OTHER_SCRIPT_RUN = r"[\u0900-\u097F\u0980-\u09FF\u0600-\u06FF\u0400-\u04FF\u0370-\u03FF\u0590-\u05FF\u0E00-\u0E7F\uAC00-\uD7AF\u3040-\u309F\u30A0-\u30FF]+"
# Contracts often use standalone X/Y/Z as neutral party variables. These must never be
# translated (e.g., into 甲/乙) or dropped, otherwise cross-references break.
_VAR_MARKER = r"\b[XYZ]\b"

FREEZE_RE = re.compile(
    rf"({_TRADEMARK_TOKEN}|{_OTHER_SCRIPT_RUN}|{_URL}|{_EMAIL}|{_WIN_PATH}|{_PLACEHOLDER}|{_PERCENT_SLOT}|"
    rf"{_EN_LEGAL_REF}|{_EN_LEGAL_REF_ABBR}|{_ZH_LEGAL_REF}|{_ZH_SCHEDULE_REF}|"
    rf"{_CLAUSE_REF}|{_ENUM_NUM}|{_ENUM_ROMAN}|{_ENUM_ALPHA}|{_DOT_LEADER}|{_UNDERSCORE_LEADER}|{_DASH_LEADER}|{_VAR_MARKER})",
    flags=re.IGNORECASE,
)


@dataclass
class FreezeResult:
    text: str
    nt_map: dict[str, str] = field(default_factory=dict)


def freeze_text(text: str) -> FreezeResult:
    nt_map: dict[str, str] = {}
    next_id = 1

    def add_token(original: str) -> str:
        nonlocal next_id
        token = nt_token(next_id)
        next_id += 1
        nt_map[token] = original
        return token

    def freeze_plain(plain: str) -> str:
        return FREEZE_RE.sub(lambda m: add_token(m.group(0)), plain)

    pieces: list[str] = []
    pos = 0
    for m in ANY_SENTINEL_RE.finditer(text):
        pieces.append(freeze_plain(text[pos : m.start()]))
        pieces.append(m.group(0))
        pos = m.end()
    pieces.append(freeze_plain(text[pos:]))
    return FreezeResult(text="".join(pieces), nt_map=nt_map)


def unfreeze_text(text: str, nt_map: dict[str, str]) -> str:
    if not nt_map:
        return text

    def repl(m: re.Match[str]) -> str:
        token = m.group(0)
        return nt_map.get(token, token)

    return re.sub(r"⟦NT:\d{4}⟧", repl, text)

File: src/freezer.rs
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashMap;

use crate::ir::FreezeMaskSpan;
use crate::sentinels::{nt_token, ANY_SENTINEL_RE, NT_RE};

#[derive(Debug, Clone)]
pub struct FreezeResult {
    pub text: String,
    pub nt_map: HashMap<String, String>,
    pub mask: Vec<FreezeMaskSpan>,
}

static FREEZE_RE: Lazy<Regex> = Lazy::new(|| {
    let url = r"https?://[^\s<>()]+";
    let email = r"[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}";
    let win_path = r#"(?:[A-Za-z]:\\(?:[^\\/:*?"<>|\r\n]+\\)*[^\\/:*?"<>|\r\n]*)"#;
    let placeholder = r"(?:\{[^{}\r\n]{1,100}\}|\$\{[^{}\r\n]{1,100}\})";
    let percent_slot = r"%\d+";
    // Preserve structured identifiers such as "4.1(b)", "1-2.3", "(i)/(ii)", and "IV".
    let clause_ref = r"(?:\b\d+(?:[.,]\d+)+(?:-\d+(?:[.,]\d+)*)?(?:\([A-Za-z0-9]+\))*\b|\b\d+(?:-\d+)+(?:\([A-Za-z0-9]+\))*\b|\b\d+\([A-Za-z0-9]+\)(?:\([A-Za-z0-9]+\))*\b|\b[IVXLCDM]{1,8}\b)";
    let enum_num = r"\(\d{1,3}\)";
    let enum_roman = r"\((?:[ivxlcdmIVXLCDM]{1,6})\)";
    let enum_alpha = r"\([A-Za-z]\)";
    let dot_leader = r"[.\u2026]{8,}";
    let underscore_leader = r"_{5,}";
    let dash_leader = r"[-\u2010\u2011\u2012\u2013\u2014\u2015\u2212]{5,}";
    let trademark_token = r"[\p{L}\p{N}]{2,24}[\u00AE\u2122\u00A9]";
    let other_script_run =
        r"[\u0900-\u097F\u0980-\u09FF\u0600-\u06FF\u0400-\u04FF\u0370-\u03FF\u0590-\u05FF\u0E00-\u0E7F\uAC00-\uD7AF\u3040-\u309F\u30A0-\u30FF]+";
    let var_marker = r"\b[XYZ]\b";

    let pat = format!(
        "({trademark_token}|{other_script_run}|{url}|{email}|{win_path}|{placeholder}|{percent_slot}|{clause_ref}|{enum_num}|{enum_roman}|{enum_alpha}|{dot_leader}|{underscore_leader}|{dash_leader}|{var_marker})"
    );
    Regex::new(&pat).expect("freeze regex")
});

pub fn freeze_text(text: &str) -> FreezeResult {
    let mut nt_map: HashMap<String, String> = HashMap::new();
    let mut mask: Vec<FreezeMaskSpan> = Vec::new();
    let mut next_id: usize = 1;

    let mut add_token = |original: &str| -> String {
        let token = nt_token(next_id);
        next_id += 1;
        nt_map.insert(token.clone(), original.to_string());
        token
    };

    let freeze_plain = |plain: &str,
                        base: usize,
                        mask: &mut Vec<FreezeMaskSpan>,
                        add_token: &mut dyn FnMut(&str) -> String|
     -> String {
        if plain.is_empty() {
            return String::new();
        }
        let mut out = String::with_capacity(plain.len());
        let mut pos = 0usize;
        for m in FREEZE_RE.find_iter(plain) {
            if m.start() > pos {
                out.push_str(&plain[pos..m.start()]);
            }
            let original = &plain[m.start()..m.end()];
            let token = add_token(original);
            mask.push(FreezeMaskSpan {
                src_start: base.saturating_add(m.start()),
                src_end: base.saturating_add(m.end()),
                token: token.clone(),
                original: original.to_string(),
            });
            out.push_str(&token);
            pos = m.end();
        }
        if pos < plain.len() {
            out.push_str(&plain[pos..]);
        }
        out
    };

    let mut pieces: Vec<String> = Vec::new();
    let mut pos = 0usize;
    for m in ANY_SENTINEL_RE.find_iter(text) {
        pieces.push(freeze_plain(
            &text[pos..m.start()],
            pos,
            &mut mask,
            &mut add_token,
        ));
        pieces.push(m.as_str().to_string());
        pos = m.end();
    }
    pieces.push(freeze_plain(&text[pos..], pos, &mut mask, &mut add_token));

    FreezeResult {
        text: pieces.concat(),
        nt_map,
        mask,
    }
}

pub fn unfreeze_text(text: &str, nt_map: &HashMap<String, String>) -> String {
    if nt_map.is_empty() || text.is_empty() {
        return text.to_string();
    }
    NT_RE
        .replace_all(text, |caps: &regex::Captures<'_>| {
            let tok = caps.get(0).unwrap().as_str();
            nt_map.get(tok).cloned().unwrap_or_else(|| tok.to_string())
        })
        .into_owned()
}
